Docs: Mutual Auth Topic page (#2820)
pomerium/pomerium: docs/docs/topics/mutual-auth.md

1. When the user connects to the route `service.example.com` the traffic is received by the Pomerium Proxy service.
1. The proxy redirects to the Authenticate service to validate the user by having them sign in with the identity provider (**IdP**).
1. After authentication succeeds, the proxy verifies that the authenticated (**authn**) user is authorized (**authz**) to access the service, and begins communication with it.

This is great, only the users that are supposed to have access to the service get through. But this model is dependent on the security of your network perimeter. If a bad actor gains access to your local area network (**LAN**), they can now communicate with the service directly:

```mermaid
flowchart TD
    D[End User]
    D <--> A
    E[/Hacker/]
    E<--->C
    subgraph LAN
    style LAN stroke-dasharray: 5 5
    A(Pomerium Proxy Service)
    B(Pomerium Authenticate Service)
    A <-.-> B
    C(Service)
    A <--> C
    end
```

While your network *should* be secured to only allow traffic at specified ports and directed to specified services, this creates a single point of failure. A hacker need only bypass your firewall to get direct access to your service.

> ## JWT Verification: Internal Mutual Auth

Reviewer: ```suggestion
## JWT Verification: Application-based Mutual Authentication
```

1. After authentication succeeds, the proxy verifies that the authenticated user is authorized to access the service, and sends the connection to Grafana, including the `X-Pomerium-JWT-Assertion` header containing the JWT signed by the key set in the Pomerium Authenticate service.
1. The service (Grafana) reads the signing key from the Authenticate service, and uses it to validate the JWT. Finding it to match, the traffic is allowed.

Our hacker may be able to forge a basic JWT, but they can't sign it with the key secured in the Pomerium configuration, so Grafana rejects their connections:

```mermaid
flowchart TD
    D[End User]
    E[/Hacker/]
    E---xC
    D<-->A
    subgraph LAN
    style LAN stroke-dasharray: 5 5
        subgraph Pomerium[ ];
        style Pomerium stroke-width: 0
            A(Pomerium Proxy Service)
            B(Pomerium Authenticate Service)
        end
            C(Grafana)
    end
    A <-.-> B
    A <--> C
    C -.-> B
```

In this way, we've applied a zero-trust security model to the application layer of our infrastructure's network model'.

> ## mTLS

Reviewer: ```suggestion
## Mutual TLS (mTLS)
```

> ## mTLS

Reviewer: ```suggestion
## mTLS: Protocol-based Mutual Authentication
```
or  (trying to make it match the jwt one)

---
title: Mutual Authentication
lang: en-US
sidebarDepth: 1
meta:
  - name: keywords
    content: pomerium identity-access-proxy mutual authentication jwt jwks mtls
description: >-
  This page describes the concept of mutual authentication and why it's important.
---

# Mutual Authentication: A Component of Zero-Trust

> Pomerium provides a good layer of security out of the box, but it's not (and can't be) configured for complete [zero trust] right out of the box. This page explains several methods of achieving mutual authentication, a big part of the zero-trust model, with practical examples.

Reviewer: ```suggestion
Pomerium provides a good layer of security out of the box, but it's not (and can't be) configured for complete [zero trust] right out of the box. This page explains several methods of achieving mutual authentication — a big part of the zero-trust model — with practical examples.
```

> This is a long page that dives into several specific security practices that provide mutual authentication. You can use the table of contents below to narrow down to the specific tools you're interested in, or read the entire doc for a deeper understanding of how these tools work together to support strong infrastructure security.

Reviewer: ```suggestion
This is a nuanced topic that dives into several specific security practices that provide mutual authentication. You can use the table of contents below to narrow down to the specific tools you're interested in or read the entire doc for a deeper understanding of how these tools work together to support strong infrastructure security.
```

[[toc]]

## What Is Mutual Authentication?

> The concept of mutual authentication is deceptively simple. It means that both sides of a connection can validate the identity of the other. The application of this concept however, can be varied and complex. Pomerium provides the features and capability to provide mutual authentication between itself and end users, as well, as between itself and upstream services, but configuring those external endpoints can vary depending on each service's features.

Reviewer: ```suggestion
The concept of mutual authentication is deceptively simple. It means that both sides of a connection can validate the identity of the other. The application of this concept, however, can be varied and complex. Pomerium provides the features and capability to provide mutual authentication between itself and end users, as well as between itself and upstream services, but configuring those external endpoints can vary depending on each service's features.
```

## Pomerium Defaults

> - **Security**: Moderate
> - **Difficulty**: Easy
> - **Encrypted**: Yes (from the end user to Pomerium. It's up to the service to provide a TLS endpoint for Pomerium to use.)
> - **Mutual Authentication**: None

> Let's look at a basic installation of Pomerium on a local network, with a single downstream service. This service contains sensitive data that we only want the right people to access.

Reviewer: ```suggestion
Let's look at a basic installation of Pomerium on a local network, with a single upstream service. This service contains sensitive data that we only want the right people to access.
```

We should add a terminology page like envoy. It's so confusing.

https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/intro/terminology

Author: I can do that.

Reviewer: @alexfornuto  should we track that effort separately? I'm sure there's other terminology that would fit the bill. Some that come to mind are:

- upstream / downstream
- identity provider
- secure enclave / TEE / security keys
- peremeter / zero trust
- least user privilege
- Namespaces
- mutual authentication
- authentication
- authorization

and so on. Even if they are just links to other resources and our own docs.

Author: Yea, I'm thinking of a separate page, linking the first instance of each term in a page to it with anchors. 

In a future iteration of docs, we can make reusable tooltips to define concepts. See [this example](https://pantheon.io/docs/style-guide#tooltips).

> Let's look at a basic installation of Pomerium on a local network, with a single downstream service. This service contains sensitive data that we only want the right people to access.

Reviewer: ```suggestion
Let's look at a basic installation of Pomerium on a local network, with a single downstream service. This service contains sensitive data that we want to ensure is confidential and cannot be tampered with.
```

::: tip Note

> This section describes multiple services of Pomerium like Authenticate and Proxy as separate services. If you are running Pomerium in all-in-one mode, these will be a single process/container in your environment.

Reviewer: See my combined PR suggestion.

:::

```mermaid
flowchart TD
    D[End User]
        D <--> A

>     subgraph LAN

Reviewer: ```suggestion
    subgraph LAN
```

Can we update all the the diagrams to call this "internal network" vs LAN?

    style LAN stroke-dasharray: 5 5
    A(Pomerium Proxy Service)
    B(Pomerium Authenticate Service)
    A <-.-> B
    C(Service)
    A <--> C
    end
```

1. When the user connects to the route `service.example.com` the traffic is received by the Pomerium Proxy service.
1. The proxy redirects to the Authenticate service to validate the user by having them sign in with the identity provider (**IdP**).
1. After authentication succeeds, the proxy verifies that the authenticated (**authn**) user is authorized (**authz**) to access the service, and begins communication with it.

This is a good start, only the users that are supposed to have access to the service get through. But this model is dependent on the security of your network perimeter. If a bad actor gains access to your local area network (**LAN**), they can now communicate with the service directly:

```mermaid
flowchart TD
    D[End User]
    D <--> A
    E[/Hacker/]
    E<--->C
    subgraph LAN
    style LAN stroke-dasharray: 5 5
    A(Pomerium Proxy Service)
    B(Pomerium Authenticate Service)
    A <-.-> B
    C(Service)
    A <--> C
    end
```

> While your network *should* be secured to only allow traffic at specified ports and directed to specified services, this creates a single point of failure. A hacker need only bypass your firewall to get direct access to your service.

Reviewer: >> A hacker **need** only bypass your firewall to get direct access to your service.

I think the verb is wrong here.

Author: > However, in a rather formal style of English, there also exists an auxiliary verb ("helping verb") need, which is followed by a bare infinitive, and is only used in negative polarity contexts: with not, with only, in questions, and so on. Additionally, it only exists in the present tense. Aside from these two restrictions, it's very much like other auxiliary verbs (can, may, should, etc.): "We need not discuss it."
> 
> So your first example ("[…] one need only look […]") is correct — using the auxiliary verb need — whereas your second example ("[…] one needs only look […]") is wrong — it should either be "one need only look" (using the auxiliary verb), or "one only needs to look" (using the regular verb).

From [english.stackexchange](https://english.stackexchange.com/questions/297233/is-one-needs-only-or-one-need-only-correct).

Reviewer: TIL thanks for the explanation.

## JWT Verification: Application-based Mutual Authentication

> - **Security**: Strong
> - **Difficulty**: Moderate
> - **Encrypted**: Yes (from the end user to Pomerium. It's up to the service to provide a TLS endpoint for Pomerium to use.)
> - **Mutual Authentication**: Application Layer

> Many, but not all, modern web applications support json web tokens (**JWTs**). These tokens are provided by Pomerium (with the [`pass_identity_headers`] key) to the downstream service so that it can independently verify that the traffic it receives is authorized.

Reviewer: ```suggestion
Many, but not all, modern web applications support [json web tokens](https://datatracker.ietf.org/doc/html/rfc7519) (**JWTs**). These tokens can be provided by Pomerium (with the [`pass_identity_headers`] key) so that an upstream service can independently verify that the traffic has properly gone through Pomerium, and that the incoming user request is authorized.
```

Let's look at an example, modeled from our [Grafana] integration guide:

```mermaid
flowchart TD
    D[End User]
    D<-->A
    subgraph LAN
    style LAN stroke-dasharray: 5 5
        subgraph Pomerium [ ];
        style Pomerium stroke-width: 0
            A(Pomerium Proxy Service)
            B(Pomerium Authenticate Service)
        end
            C(Grafana)
    end
    A <-.-> B
    A <--> C
    C -.-> B
```

> 1. When the user connects to the route `grafana.example.com` the traffic is received by the Pomerium Proxy service.

Reviewer: ```suggestion
1. When the user connects to the route `grafana.example.com` the traffic is received by the `Pomerium Proxy` service.
```

> 1. The proxy redirects to the Authenticate service to validate the user by having them sign in with the identity provider (**IdP**).

Reviewer: ```suggestion
1. The proxy redirects to the `Pomerium Authenticate` service to validate the user by having them sign in with the identity provider (**IdP**).
```

Author: I get the addition of Pomerium, but why make it an inline code snippet?

1. After authentication succeeds, the proxy verifies that the authenticated user is authorized to access the service, and sends the connection to Grafana, including the `X-Pomerium-JWT-Assertion` header containing the JWT signed by the key set in the Pomerium Authenticate service.
1. The service (Grafana) reads the signing key from the Authenticate service, and uses it to validate the JWT. Finding it to match, the traffic is allowed.

Our hacker may be able to forge a basic JWT, but they can't sign it with the key secured in the Pomerium configuration, so Grafana rejects their connections:

```mermaid
flowchart TD
    D[End User]
    E[/Hacker/]
    E---xC
    D<-->A
    subgraph LAN
    style LAN stroke-dasharray: 5 5
        subgraph Pomerium[ ];
        style Pomerium stroke-width: 0
            A(Pomerium Proxy Service)
            B(Pomerium Authenticate Service)
        end

>             C(Grafana)

Reviewer: ```suggestion
            C(Grafana)
```

It's confusing to jump between examples because in some places the upstream service is called `Service` and in others `Grafana`. We should be consistent.

    end
    A <-.-> B
    A <--> C
    C -.-> B
```

In this way, we've applied a zero-trust security model to the application layer of our infrastructure's network model.

## mTLS: Protocol-based Mutual Authentication

> - **Security**: Excellent
> - **Difficulty**: Hard
> - **Encrypted**: Yes
> - **Mutual Authentication**: Protocol Layer

Most tech professionals are familiar with [Transport Layer Security] (**TLS**). The majority of traffic on the web today is sent using TLS. In addition to encrypting data using the server's TLS certificate, the server identity is validated by the certificate and the Certificate Authority (**CA**) that signed it.

```mermaid
flowchart RL
    subgraph pc [PC]
    D(Browser) -.-> B(Trusted Keystore)
    end
    subgraph Server
    D-. http .-> A[Service]
    A-. Certificate .->D
    end
    D<-- https -->A
```

1. The browser initiates a connection to `example.com` over the `http` protocol.
1. The server sends its public certificate to the browser.

> 1. The browser reads the certificate chain to find the CA, and checks against the computers keystore to see if the CA is one that it trusts.

Reviewer: ```suggestion
1. The browser reads the certificate chain to find the CA, and checks against the computer's keystore to see if the CA is one that it trusts.
```

1. After confirming the CA is trusted the browser reconnects to the server, this time using the `https` protocol and encrypting the traffic using the public certificate.

The process above confirms the identity of the *server*, protecting the client. Mutual TLS (**mTLS**) allows the server to confirm the identity of the *client* using a client certificate.

```mermaid
flowchart RL
    subgraph Server
    A[Service]<-.->C[Trusted Keystore]
    end
    D(Browser)-- Client Certificate -->A
    A-- Client Certificate Request --->D
```

::: tip
The exchange diagrammed here occurs *after* the initial connection and confirmation of the server-side certificate shown above.
:::

1. After the server certificate is trusted and an `HTTPS` connection is established, the server requests a client certificate.

> 1. The user is prompted to use one of the certificates previously imported into thr browser. This certificate is send to the server

Reviewer: ```suggestion
1. The user is prompted to use one of the certificates previously imported into the browser. This certificate is sent to the server
```

## Pomerium Defaults

> - **Security**: Moderate
> - **Difficulty**: Easy
> - **Encrypted**: Yes (from the end user to Pomerium. It's up to the service to provide a TLS endpoint for Pomerium to use.)
> - **Mutual Authentication**: None

Let's look at a basic installation of Pomerium on a local network, with a single upstream service. This service contains sensitive data that we want to ensure is confidential and cannot be tampered with.

::: tip Note
Throughout this guide we will diagram Pomerium as a single service, as it is in all-in-one mode. This is the version provided by our [binaries] and in our Docker-based [Quick Start] page.
:::

```mermaid
flowchart LR
  C(End User)
  D(Identity Provider)
  C==o A
  A-.-D
  subgraph lan[Internal Network]
    style lan stroke-dasharray: 5 5
    A(Pomerium)
    B(Upstream Service)
    A --- B
  end
```

1. When the user connects to the route `service.example.com` the traffic is received by Pomerium.
1. Pomerium redirects the user to the identity provider (**IdP**) to sign in and validate their identity.

> 1. After signing in to the IdP, Pomerium verifies that the authenticated (**authn**) user is authorized (**authz**) to access the service, and begins communication with it.

Reviewer: ```suggestion
1. After signing in to the IdP, Pomerium verifies that the authenticated (**AuthN**) user is authorized (**AuthZ**) to access the service, and begins communication with it.
```

More of a convention but I typically see these mixed cased. There may be other instances in this docs but I'm just flagging the first.

Reviewer: https://en.wikipedia.org/wiki/Web_API_security

> Along with the ease of API integrations come the difficulties of ensuring proper authentication (AuthN) and authorization (AuthZ).

This is a good start, only the users that are supposed to have access to the service get through. But this model is dependent on the security of your network perimeter. If a bad actor gains access to your internal network, they can now communicate with the service directly:

```mermaid
flowchart LR
  C(End User)
  C==o A
  subgraph lan[Internal Network]
    style lan stroke-dasharray: 5 5
    A(Pomerium)
    B(Upstream Service)
    A --- B
  end
  E[/Hacker/]---B
```

While your network *should* be secured to only allow traffic at specified ports and directed to specified services, this creates a single point of failure. A hacker need only bypass your firewall to get direct access to your service.

## JWT Verification: Application-based Mutual Authentication

> - **Security**: Strong
> - **Difficulty**: Moderate
> - **Encrypted**: Yes (from the end user to Pomerium. It's up to the service to provide a TLS endpoint for Pomerium to use.)
> - **Mutual Authentication**: Application Layer

> Many, but not all, modern web applications support [json web tokens][jwt-rfc] (**JWTs**). These tokens can be provided by Pomerium (with the [`pass_identity_headers`] key) so that an upstream service can independently verify that the traffic has properly gone through Pomerium, and that the incoming user request is authorized.

Reviewer: ```suggestion
Many, but not all, modern web applications support [JSON web tokens][jwt-rfc] (**JWTs**). These tokens can be provided by Pomerium (with the [`pass_identity_headers`] key) so that an upstream service can independently verify that the traffic has properly gone through Pomerium, and that the incoming user request is authorized.
```

I believe JSON should be capitalized. 

https://en.wikipedia.org/wiki/JSON

Reviewer: There may be other instances I missed.

1. The upstream service reads the signing key from Pomerium, and uses it to validate the JWT. Finding that the JWT was signed by the trusted key, the traffic is allowed.

Our hacker may be able to forge a basic JWT, but they can't sign it with the key secured in the Pomerium configuration, so the upstream service rejects their connections:

```mermaid
flowchart LR
D(End User)
C(Identity Provider)
D==oA
subgraph lan[Internal Network]
  direction LR
  style lan stroke-dasharray: 5 5
  A(Pomerium)
  B(Upstream Service)
  A ==o B
end
B-.JWKS Verification.-A
C-.-A
E[/Hacker/] --x B
```

In this way, we've applied a zero-trust security model to the application layer of our infrastructure's network model. You can see JWT verification in practice with our [Grafana] integration guide.

## mTLS: Protocol-based Mutual Authentication

> - **Security**: Excellent
> - **Difficulty**: Hard
> - **Encrypted**: Yes
> - **Mutual Authentication**: Protocol Layer

> Most tech professionals are familiar with [Transport Layer Security] (**TLS**). The majority of traffic on the web today is sent using TLS. In addition to encrypting data using the server's TLS certificate, the server identity is validated by the certificate and the Certificate Authority (**CA**) that signed it.

Reviewer: Maybe we should add a subheading clarifying mTLS between browser and pomerium and pomerium and upstreams?

Author: Resolved by review changes.

> Most tech professionals are familiar with [Transport Layer Security] (**TLS**). The majority of traffic on the web today is sent using TLS. In addition to encrypting data using the server's TLS certificate, the server identity is validated by the certificate and the Certificate Authority (**CA**) that signed it.

Reviewer: ```suggestion
### [North-south mTLS](https://en.wikipedia.org/wiki/North-south_traffic)

Most tech professionals are familiar with [Transport Layer Security] (**TLS**). The majority of traffic on the web today is sent using TLS. In addition to encrypting data using the server's TLS certificate, the server identity is validated by the certificate and the Certificate Authority (**CA**) that signed it.
```

1. After confirming the CA is trusted the browser reconnects to the server, this time using the `https` protocol and encrypting the traffic using the public certificate.

The process above confirms the identity of the *server*, protecting the client. Mutual TLS (**mTLS**) allows the server to confirm the identity of the *client* using a client certificate.

```mermaid
flowchart RL
  subgraph pc [PC]
    D(Browser) -.- B(Trusted Keystore)
  end
  subgraph Server
    A(Service)-.-C(Trusted Keystore)
  end
  D(Browser)-. Client Certificate .->A
  A-. Server Certificate ..->D
  D<==>A
  style pc fill: white, stroke: black
  style Server fill: white, stroke: black
```

1. After the server certificate is trusted and an `HTTPS` connection is established, the server requests a client certificate.
1. The user is prompted to use one of the certificates previously imported into the browser. This certificate is sent to the server
1. The server validates the client certificate signing authority against its trusted keystore or authorized client CA. Once authorized, the server resumes normal encrypted communication with the client.

> 

Reviewer: ```suggestion
### [East-west mTLS](https://en.wikipedia.org/wiki/East-west_traffic)
```

https://en.wikipedia.org/wiki/East-west_traffic

Reviewer: Though we should probably add this to the terminology doc too.

  B-.JWT.-D
  B<==mTLS==>E
  B<==mTLS===>F
  G<==mTLS==>B
  C-.-B
```

- In this example, End users authenticate with Pomerium to access the user app:
  - Their connection to Pomerium is authenticated after signing in to their identity provider, and encrypted with TLS.
  - The communication between Pomerium and the user app is mutually authenticated with mTLS, and the user's JWT is passed from Pomerium to the app to confirm the user identity to the service. The service validates the JWT against Pomerium's signing Key.
- The admin also authenticates with Pomerium to access their services:
  - The connection is authenticated and encrypted the same as the user's, but the route to the admin app also requires a client certificate for mutual authentication at the protocol layer.
  - The connections from the admin to the admin services are now mutually authenticated both upstream and downstream from Pomerium.
- The API server is required by both the user and admin apps. Rather than build authentication into the API service, it is accessible only through a sidecar. The sidecar only accepts mTLS-authenticated connections from the proxy, so both apps connect to the API through Pomerium.

> ## Legend

Reviewer: Could we move this to the top of the doc?

Author: Discussed in Slack.
